File: jscomp/test/queue_402.ml
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*        Francois Pottier, projet Cristal, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 2002 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(***********************************************************************)

exception Empty

(* OCaml currently does not allow the components of a sum type to be
   mutable. Yet, for optimal space efficiency, we must have cons cells
   whose [next] field is mutable. This leads us to define a type of
   cyclic lists, so as to eliminate the [Nil] case and the sum
   type. *)

type 'a cell = {
    content: 'a;
    mutable next: 'a cell
  }

(* A queue is a reference to either nothing or some cell of a cyclic
   list. By convention, that cell is to be viewed as the last cell in
   the queue. The first cell in the queue is then found in constant
   time: it is the next cell in the cyclic list. The queue's length is
   also recorded, so as to make [length] a constant-time operation.

   The [tail] field should really be of type ['a cell option], but
   then it would be [None] when [length] is 0 and [Some] otherwise,
   leading to redundant memory allocation and accesses. We avoid this
   overhead by filling [tail] with a dummy value when [length] is 0.
   Of course, this requires bending the type system's arm slightly,
   because it does not have dependent sums. *)

type 'a t = {
    mutable length: int;
    mutable tail: 'a cell
  }

let create () = {
  length = 0;
  tail = Obj.magic None
}

let clear q =
  q.length <- 0;
  q.tail <- Obj.magic None

let add x q =
  if q.length = 0 then
    let rec cell = {
      content = x;
      next = cell
    } in
    q.length <- 1;
    q.tail <- cell
  else
    let tail = q.tail in
    let head = tail.next in
    let cell = {
      content = x;
      next = head
    } in
    q.length <- q.length + 1;
    tail.next <- cell;
    q.tail <- cell

let push =
  add

let peek q =
  if q.length = 0 then
    raise Empty
  else
    q.tail.next.content

let top =
  peek

let take q =
  if q.length = 0 then raise Empty;
  q.length <- q.length - 1;
  let tail = q.tail in
  let head = tail.next in
  if head == tail then
    q.tail <- Obj.magic None
  else
    tail.next <- head.next;
  head.content

let pop =
  take

let copy q =
  if q.length = 0 then
    create()
  else
    let tail = q.tail in

    let rec tail' = {
      content = tail.content;
      next = tail'
    } in

    let rec copy prev cell =
      if cell != tail
      then let res = {
        content = cell.content;
        next = tail'
      } in prev.next <- res;
      copy res cell.next in

    copy tail' tail.next;
    {
      length = q.length;
      tail = tail'
    }

let is_empty q =
  q.length = 0

let length q =
  q.length

let iter f q =
  if q.length > 0 then
    let tail = q.tail in
    let rec iter cell =
      f cell.content;
      if cell != tail then
        iter cell.next in
    iter tail.next

let fold f accu q =
  if q.length = 0 then
    accu
  else
    let tail = q.tail in
    let rec fold accu cell =
      let accu = f accu cell.content in
      if cell == tail then
        accu
      else
        fold accu cell.next in
    fold accu tail.next

let transfer q1 q2 =
  let length1 = q1.length in
  if length1 > 0 then
    let tail1 = q1.tail in
    clear q1;
    if q2.length > 0 then begin
      let tail2 = q2.tail in
      let head1 = tail1.next in
      let head2 = tail2.next in
      tail1.next <- head2;
      tail2.next <- head1
    end;
    q2.length <- q2.length + length1;
    q2.tail <- tail1

File: jscomp/test/queue_402.js
'use strict';

var Curry = require("../../lib/js/curry.js");
var Caml_obj = require("../../lib/js/caml_obj.js");
var Caml_exceptions = require("../../lib/js/caml_exceptions.js");

var Empty = Caml_exceptions.create("Queue_402.Empty");

function create(param) {
  return {
          length: 0,
          tail: undefined
        };
}

function clear(q) {
  q.length = 0;
  q.tail = undefined;
  return /* () */0;
}

function add(x, q) {
  if (q.length === 0) {
    var cell = { };
    cell.content = x;
    cell.next = cell;
    q.length = 1;
    q.tail = cell;
    return /* () */0;
  } else {
    var tail = q.tail;
    var head = tail.next;
    var cell$1 = {
      content: x,
      next: head
    };
    q.length = q.length + 1 | 0;
    tail.next = cell$1;
    q.tail = cell$1;
    return /* () */0;
  }
}

function peek(q) {
  if (q.length === 0) {
    throw Empty;
  }
  return q.tail.next.content;
}

function take(q) {
  if (q.length === 0) {
    throw Empty;
  }
  q.length = q.length - 1 | 0;
  var tail = q.tail;
  var head = tail.next;
  if (head === tail) {
    q.tail = undefined;
  } else {
    tail.next = head.next;
  }
  return head.content;
}

function copy(q) {
  if (q.length === 0) {
    return {
            length: 0,
            tail: undefined
          };
  } else {
    var tail = q.tail;
    var tail$prime = { };
    Caml_obj.caml_update_dummy(tail$prime, {
          content: tail.content,
          next: tail$prime
        });
    var copy$1 = function (_prev, _cell) {
      while(true) {
        var cell = _cell;
        var prev = _prev;
        if (cell !== tail) {
          var res = {
            content: cell.content,
            next: tail$prime
          };
          prev.next = res;
          _cell = cell.next;
          _prev = res;
          continue ;
        } else {
          return /* () */0;
        }
      };
    };
    copy$1(tail$prime, tail.next);
    return {
            length: q.length,
            tail: tail$prime
          };
  }
}

function is_empty(q) {
  return q.length === 0;
}

function length(q) {
  return q.length;
}

function iter(f, q) {
  if (q.length > 0) {
    var tail = q.tail;
    var _cell = tail.next;
    while(true) {
      var cell = _cell;
      Curry._1(f, cell.content);
      if (cell !== tail) {
        _cell = cell.next;
        continue ;
      } else {
        return /* () */0;
      }
    };
  } else {
    return /* () */0;
  }
}

function fold(f, accu, q) {
  if (q.length === 0) {
    return accu;
  } else {
    var tail = q.tail;
    var _accu = accu;
    var _cell = tail.next;
    while(true) {
      var cell = _cell;
      var accu$1 = _accu;
      var accu$2 = Curry._2(f, accu$1, cell.content);
      if (cell === tail) {
        return accu$2;
      } else {
        _cell = cell.next;
        _accu = accu$2;
        continue ;
      }
    };
  }
}

function transfer(q1, q2) {
  var length1 = q1.length;
  if (length1 > 0) {
    var tail1 = q1.tail;
    clear(q1);
    if (q2.length > 0) {
      var tail2 = q2.tail;
      var head1 = tail1.next;
      var head2 = tail2.next;
      tail1.next = head2;
      tail2.next = head1;
    }
    q2.length = q2.length + length1 | 0;
    q2.tail = tail1;
    return /* () */0;
  } else {
    return /* () */0;
  }
}

var push = add;

var top = peek;

var pop = take;

exports.Empty = Empty;
exports.create = create;
exports.clear = clear;
exports.add = add;
exports.push = push;
exports.peek = peek;
exports.top = top;
exports.take = take;
exports.pop = pop;
exports.copy = copy;
exports.is_empty = is_empty;
exports.length = length;
exports.iter = iter;
exports.fold = fold;
exports.transfer = transfer;
/* No side effect */
